File: src/waypoint/sources/windows_update.py
"""Windows Update Catalog driver source.

Uses the same official channel Windows Update itself uses for driver
offers — `Microsoft.Update.Session` / `IUpdateSearcher`, queried for
`IsInstalled=0 and Type='Driver'` — rather than a third-party aggregator.
References:
  https://learn.microsoft.com/en-ca/answers/questions/5656657/microsoft-update-catalog-searching-for-firmware-up
  https://learn.microsoft.com/en-us/windows/win32/api/wuapi/nf-wuapi-iupdatesearcher-search

Not yet validated against real hardware — first implementation pass, to be
exercised on the Dell/Asus dev machines before being trusted the way
Meridian OS components are only marked "working" after bare-metal
validation.
"""

from __future__ import annotations

from datetime import date

from waypoint.core.models import DriverCandidate, SignatureType


class WindowsUpdateCatalogSource:
    source_id = "windows_update_catalog"

    def search(self, hwids: list[str]) -> list[DriverCandidate]:
        # Lazily imported: COM automation only works on Windows.
        import win32com.client  # type: ignore

        session = win32com.client.Dispatch("Microsoft.Update.Session")
        searcher = session.CreateUpdateSearcher()
        search_result = searcher.Search("IsInstalled=0 and Type='Driver'")

        candidates: list[DriverCandidate] = []
        hwid_set = {h.lower() for h in hwids}
        for update in search_result.Updates:
            driver_hwid = getattr(update, "DriverHardwareID", "") or ""
            if driver_hwid.lower() not in hwid_set:
                continue
            driver_date = None
            raw_date = getattr(update, "DriverVerDate", None)
            if raw_date:
                driver_date = date(raw_date.year, raw_date.month, raw_date.day)
            candidates.append(
                DriverCandidate(
                    hwid=driver_hwid,
                    class_guid=getattr(update, "DriverClass", "") or "",
                    version=getattr(update, "DriverVerVersion", "unknown"),
                    driver_date=driver_date,
                    publisher=getattr(update, "DriverManufacturer", "unknown"),
                    # Windows Update-delivered drivers are Microsoft
                    # attestation-signed at minimum; true WHQL status needs
                    # a follow-up catalog inspection call, tracked as an
                    # open item rather than assumed here.
                    signature_type=SignatureType.ATTESTATION,
                    sha256="",  # populated on fetch(), not known at search time
                    size_bytes=int(getattr(update, "MaxDownloadSize", 0) or 0),
                    source_id=self.source_id,
                    source_url=getattr(update, "SupportUrl", "") or "",
                    download_uri="",  # resolved via BITS download in fetch()
                )
            )
        return candidates

    def fetch(self, candidate: DriverCandidate, dest_dir: str) -> str:
        raise NotImplementedError(
            "Windows Update Catalog download via BITS/IUpdateDownloader is a "
            "follow-up milestone — search() is functional first, download "
            "second, matching the phased plan in docs/Architecture.md."
        )

File: rust/crates/sources/src/windows_update.rs
//! Windows Update Catalog driver source. Partial port of
//! `sources/windows_update.py`.
//!
//! Uses the same official channel Windows Update itself uses for driver
//! offers — `Microsoft.Update.Session` / `IUpdateSearcher`, queried for
//! `IsInstalled=0 and Type='Driver'` — rather than a third-party
//! aggregator. References:
//! - <https://learn.microsoft.com/en-us/windows/win32/api/wuapi/nf-wuapi-iupdatesearcher-search>
//!
//! **Porting status (honest, not a placeholder pretending otherwise):**
//! the Python version drives this API through `win32com.client`'s
//! late-bound COM automation, which auto-generates method/property
//! dispatch at runtime. Rust has no equivalent auto-dispatch — calling
//! `IUpdateSearcher::Search` and walking the resulting `IUpdateCollection`
//! requires either hand-written `IDispatch::Invoke` plumbing via the
//! `windows` crate, or a `#[interface]`-generated typelib binding for
//! `wuapi.dll`. That plumbing has not been written yet, matching the
//! Python file's own `fetch()` (already `NotImplementedError` there) —
//! `search()` here is intentionally left unimplemented too, rather than
//! faked, until the COM binding work is scoped as its own task.

use waypoint_core::DriverCandidate;

use crate::base::DriverSource;

pub struct WindowsUpdateCatalogSource;

impl WindowsUpdateCatalogSource {
    pub fn new() -> Self {
        Self
    }
}

impl Default for WindowsUpdateCatalogSource {
    fn default() -> Self {
        Self::new()
    }
}

impl DriverSource for WindowsUpdateCatalogSource {
    fn source_id(&self) -> &str {
        "windows_update_catalog"
    }

    fn search(&self, _hwids: &[String]) -> Result<Vec<DriverCandidate>, String> {
        Err("Windows Update Catalog search() needs IDispatch-based COM automation \
             bindings for Microsoft.Update.Session — not yet ported from win32com, \
             tracked as a follow-up milestone, same as fetch() below."
            .to_string())
    }

    fn fetch(&self, _candidate: &DriverCandidate, _dest_dir: &str) -> Result<String, String> {
        Err("Windows Update Catalog download via BITS/IUpdateDownloader is a \
             follow-up milestone — search() is functional first, download \
             second, matching the phased plan in docs/Architecture.md."
            .to_string())
    }
}
